### PR TITLE
DT: Add support for PI4IOE5V96248, TMP1075

### PR DESCRIPTION
**Pull request by HatsyRei**
Device tree and kernel configuration changes to support IO expander and thermal sensors.

### CODE REVIEW

File: recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch [hunk 1]
```diff
@@ -0,0 +1,262 @@
...
+ /* Talk to 8-bit I/O expander */
+ 
+-static int i2c_write_le8(struct i2c_client *client, unsigned data)
++static int i2c_write_le8(struct i2c_client *client, u64 data)
+ {
+-	return i2c_smbus_write_byte(client, data);
++	return (int)i2c_smbus_write_byte(client, data);
+ }
+ 
+-static int i2c_read_le8(struct i2c_client *client)
```
Comment (tanchunhau) [inline recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch:54]: Before that i worry, not sure after modified, 
 will it impact the IO expander in PEP (pca9674) ? 

 But I think it should works, because it still in the .compatible list after patch,
 and only remove/add static function in the gpio pcf857x.c  
 
 thanks for sharing this patch method !
Original : https://elixir.bootlin.com/linux/v4.9.12/source/drivers/gpio/gpio-pcf857x.c#L317

File: recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+@@ -74,31 +77,42 @@
+ 	struct irq_chip		irqchip;
+ 	struct i2c_client	*client;
+ 	struct mutex		lock;		/* protect 'out' */
+-	unsigned		out;		/* software latch */
+-	unsigned		status;		/* current status */
+-	unsigned		irq_enabled;	/* enabled irqs */
++	u64			out;		/* software latch */
++	u64			status;		/* current status */
++	unsigned int		irq_parent;
```
Comment (karhin1990) [inline recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch]: I believe the code that you are referring was written based older version of kernel, where the `pcf857x` struct still have `irq_parent` element. This element actually was dropped in recent kernel version (see [patch](https://git.toradex.com/cgit/linux-toradex.git/commit/drivers/gpio/gpio-pcf857x.c?h=toradex_5.4-2.3.x-imx&id=7f2f787c10596f486644d730a0a23e78abe8cbe0)) hence it was not in our gpio-pcf857x.c (our OE is pulling kernel version 'toradex_5.4-2.3.x-imx', which I believe is base off mainline kernel v5.2).

tl;dr
We should not reintroduce this element since it was removed in the kernel version that we are using.

File: recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+ 
+ /*-------------------------------------------------------------------------*/
+ 
+ /* Talk to 8-bit I/O expander */
+ 
+-static int i2c_write_le8(struct i2c_client *client, unsigned data)
++static int i2c_write_le8(struct i2c_client *client, u64 data)
+ {
+-	return i2c_smbus_write_byte(client, data);
++	return (int)i2c_smbus_write_byte(client, data);
```
Comment (karhin1990) [inline recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch]: since the code previously compiled successfully, so this change might not be needed

File: recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
++	*data = (buf[1] << 8) | buf[0];
++
++	return status;
++}
++
++/* Talk to 48-bit I/O expander */
++
++static int i2c_write_le48(struct i2c_client *client, u64 word)
++{
++	u8 buf[6] = { word & 0xff, (word >> 8) & 0xff, (word >> 16) & 0xff, (word >> 24) & 0xff, (word >> 32) & 0xff,  (word >> 40) & 0xff, };
```
Comment (karhin1990) [inline recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch]: nit: Additional space before `(word >> 40)`

File: recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
++		error = irq_set_irq_wake(gpio->irq_parent, on);
++		if (error) {
++			dev_dbg(&gpio->client->dev,
++				"irq %u doesn't support irq_set_wake\n",
++				gpio->irq_parent);
++			gpio->irq_parent = 0;
++		}
++	}
++	return error;
+ }
```
Comment (karhin1990) [inline recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch]: Same comment as the `irq_parent`. See [patch](https://git.toradex.com/cgit/linux-toradex.git/commit/drivers/gpio/gpio-pcf857x.c?h=toradex_5.4-2.3.x-imx&id=7f2f787c10596f486644d730a0a23e78abe8cbe0)

File: recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
++	if(of_property_read_string(np, "label", &gpio->chip.label))
++		gpio->chip.label = client->name;
+ 
+ 	gpio->client = client;
+ 	i2c_set_clientdata(client, gpio);
+@@ -367,6 +436,7 @@
+ 
+ 		gpiochip_set_nested_irqchip(&gpio->chip, &gpio->irqchip,
+ 					    client->irq);
++		gpio->irq_parent = client->irq;
```
Comment (karhin1990) [inline recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch]: Same comment as the `irq_parent`. See [patch](https://git.toradex.com/cgit/linux-toradex.git/commit/drivers/gpio/gpio-pcf857x.c?h=toradex_5.4-2.3.x-imx&id=7f2f787c10596f486644d730a0a23e78abe8cbe0)

File: recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch (hunk 1 above)
Comment (tanchunhau) [reply recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch:54]: @HatsyRei,
Wish to learn this more from you, 
let say the worst case, this patch has bad impact to the IO expander in PEP (pca9674).
we can have our own file (something.c) and then compile as part of kernel then load it as driver.
We can inject the file to kernal source code in linux-toradex_5.4-2.3.x.bbappend right ?
`do_configure_prepend()`

Thank you.

Comment (HatsyRei) [reply recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch:54]: @tanchunhau I think I overlooked some things when I said this is doable yesterday. If we introduce an entirely new kernel driver, then we may need to figure out how to introduce a custom kernel configuration entry (e.g. CONFIG_SOMETHING). We also need to introduce a patch to the driver Makefile to tell the kernel to build that something.c driver. I haven't figured out those parts yet, but I guess it would be a lot more work to introduce our own driver instead of patching an existing one.

Comment (tanchunhau) [reply recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch:54]: okay, thanks for the sharing.

Comment (HatsyRei) [reply recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch:54]: It might be easier than I imagine though. You can look at Kconfig and Makefile in this path:
https://git.toradex.com/cgit/linux-toradex.git/tree/drivers/gpio?h=toradex_5.4-2.3.x-imx

These are probably the files we have to patch, if we introduce a new driver.

Comment (tanchunhau) [reply recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch:54]: Thank you for sharing, I like the patching method.

File: recipes-kernel/linux/files/imx8qxp-apalis-smartracks.dtsi [hunk 7]
```diff
@@ -157,6 +157,22 @@
...
     };
+    
+    /* 48-bit IO expander on fanout board */
+    pi4ioe5v: pi4ioe5v@27 {
+        compatible = "diodes,pi4ioe5v96248";
+        reg = <0x27>;
+        gpio-controller;
+        #gpio-cells = <2>;
+        npgio=<48>;
+        gpio-line-names =
```
Comment (karhin1990) [inline recipes-kernel/linux/files/imx8qxp-apalis-smartracks.dtsi:168]: As discussed, we need to expose these lines in rcu-service so that user can access it.

**Review by karhin1990 — COMMENTED** [review]
Really good work!

File: recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch (hunk 2 above)
Comment (HatsyRei) [reply recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch]: Fixed.

File: recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch (hunk 3 above)
Comment (HatsyRei) [reply recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch]: Fixed.

File: recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch (hunk 4 above)
Comment (HatsyRei) [reply recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch]: Fixed.

File: recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch (hunk 5 above)
Comment (HatsyRei) [reply recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch]: Fixed.

File: recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch (hunk 6 above)
Comment (HatsyRei) [reply recipes-kernel/linux/files/add-support-for-diodes-pi4oe5v96248.patch]: Fixed.

File: recipes-kernel/linux/files/imx8qxp-apalis-smartracks.dtsi (hunk 7 above)
Comment (karhin1990) [reply recipes-kernel/linux/files/imx8qxp-apalis-smartracks.dtsi:168]: Marking it as resolve now. Remember to address it in another PR

**Review by karhin1990 — APPROVED** [review]

File: recipes-kernel/linux/files/imx8qxp-apalis-smartracks.dtsi [hunk 8]
```diff
@@ -157,6 +157,22 @@
...
+    pi4ioe5v: pi4ioe5v@27 {
+        compatible = "diodes,pi4ioe5v96248";
+        reg = <0x27>;
+        gpio-controller;
+        #gpio-cells = <2>;
+        npgio=<48>;
+        gpio-line-names = 
+            "IOXPD_IO0_0", "IOXPD_IO0_1", "IOXPD_IO0_2", "IOXPD_IO0_3", "IOXPD_IO0_4", "IOXPD_IO0_5", "IOXPD_IO0_6", "IOXPD_IO0_7",
+            "IOXPD_IO1_0", "IOXPD_IO1_1", "IOXPD_IO1_2", "IOXPD_IO1_3", "IOXPD_IO1_4", "IOXPD_IO1_5", "IOXPD_IO1_6", "IOXPD_IO1_7",
+            "IOXPD_IO2_0", "IOXPD_IO2_1", "IOXPD_IO2_2", "IOXPD_IO2_3", "IOXPD_IO2_4", "IOXPD_IO2_5", "IOXPD_IO2_6", "IOXPD_IO2_7",
```
Comment (tanchunhau) [inline recipes-kernel/linux/files/imx8qxp-apalis-smartracks.dtsi]: Since we have at least two IO Expander .
wish to have a GPIO name that can differentiate them, 
although we can have same name in different chip, and I can generate my old enum name to map it.
But wish my enum name is same as GPIO name is better.

wish to know your opinions, thank you.
From example：
IOXPD_IO2_0  --> IOXPD**1**_IO2_0
or 
IOXPD_IO2_0  --> IOXPD**2**_IO2_0

image

Comment (HatsyRei) [reply recipes-kernel/linux/files/imx8qxp-apalis-smartracks.dtsi]: Updated.
